woocommerce/wc-admin
# Enable Typescript checking on ./client folder

This PR aims to enable Typescript checking on ./client folder.

At time of writing, there are still many Typescript errors that need to be resolved so it has not been made mandatory to pass Typescript checks in CI.

### Detailed Change Descriptions:
- Made sub-repos composite typescript packages where necessary, as that is a requirement for enabling type checks at the parent level
- Fixed npm run ts:check command to do check
- Created a vscode task to run typescript check without failing the build process (more at: p90Yrv-2NC-p2)
- @woocommerce/explat had some type errors due to unexported types from @wordpress/explat-client-react-helpers, temporarily resolved this by using patch-package to patch it in post-install. Need to talk to @wordpress/explat-client-react-helpers team to resolve this, as I have not investigated how they can fix it properly
- changed ./client tests to use ts-jest instead of babel-jest, because babel-jest wasn't playing well with the new tsconfig
- ./packages/js-tests was changed to have separate build (for commonjs output) and type checking commands (this should be the default way that all packages are built, but we will get around to that)

## Test instructions
1. `npm install`
2. `npm run test:client`

All tests should still pass, but there will be a number of warnings about type errors.

For the soft Typescript check which just highlights type errors in the IDE refer to p90Yrv-2NC-p2.

To run the Typescript check manually, run `npm run ts:check` or `npm run ts:check:watch`

There should be no changes to the application or any build process output.

## Related:

woocommerce/woocommerce#32179
woocommerce/woocommerce-admin#8282
